src/main/java/com/powsybl/openloadflow/network/util/ActivePowerDistribution.java

```diff
@@ -104,4 +108,60 @@ public static Step getStep(LoadFlowParameters.BalanceType balanceType, boolean l
                     new GenerationActivePowerDistributionStep(GenerationActivePowerDistributionStep.ParticipationType.REMAINING_MARGIN, useActiveLimits);
         };
     }
+
+    public record ResultAfterFailureBehaviorHandling(boolean failed, String failedMessage, int iteration, double remainingMismatch, boolean movedBuses, double additionalDistributedActivePower) { }
+
+    public static ResultAfterFailureBehaviorHandling handleDistributionFailureBehavior(OpenLoadFlowParameters.SlackDistributionFailureBehavior behavior,
+                                                                                       LfGenerator referenceGenerator,
+                                                                                       double requestedActivePower,
+                                                                                       Result result, String failMessageTemplate) {
+        Objects.requireNonNull(behavior);
+        Objects.requireNonNull(result);
+        ResultAfterFailureBehaviorHandling resultAfterFailureBehaviorHandling;
+        final OpenLoadFlowParameters.SlackDistributionFailureBehavior effectiveBehavior;
+
+        // if requested behavior is to distribute on reference generator, but there is no reference generation, we fall back internally to FAIL mode
+        if (OpenLoadFlowParameters.SlackDistributionFailureBehavior.DISTRIBUTE_ON_REFERENCE_GENERATOR == behavior && referenceGenerator == null) {
+            effectiveBehavior = OpenLoadFlowParameters.SlackDistributionFailureBehavior.FAIL;
```

Comment:
A log message may be given to inform that no reference generator is defined.

Reply:
adding it but proposing DEBUG level, I fear this would be too noisy: it would be issued at each outerloop